xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+			// the current LB policies perspective. Also, the current LB is
+			// still being used by grpc to choose SubConns per RPC, and thus
+			// should use the most updated form of the current balancer.
+			gsb.cc.UpdateState(state)
+		}
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	finished := make(chan struct{})
+	go func() {
+		gsb.swapMu.Lock()
+		defer gsb.swapMu.Unlock()
+		gsb.balRecentlyClosed = gsb.balancerCurrent
+		gsb.balancerCurrent.Close()
+		for sc, bal := range gsb.scToSubBalancer {
+			if bal == gsb.balancerCurrent {
+				gsb.cc.RemoveSubConn(sc) // Or does this happen implicitly?
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+
+		gsb.balancerCurrent = gsb.balancerPending
+		gsb.balancerPending = nil
+		finished <- struct{}{}
+	}()
+	<-finished
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bal balancer.Balancer) bool {
+	return bal == gsb.balancerCurrent || bal == gsb.balancerPending
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if gsb.closed.HasFired() {
+		// logger?
+		return nil, errBalancerClosed
+	}
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bal, bal)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+// Eat newSubConn calls from current if there is a pending?
+
+// Eat NewAddresses calls from current if there is a pending? (Allowing ClientConn to do work that doesn't need)
+// close guard for the three functions below?
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		return
+	}
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) removeSubConn(bal balancer.Balancer, sc balancer.SubConn) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if bal != gsb.balancerCurrent && bal != gsb.balancerPending {
+		return
+	}
+	gsb.cc.RemoveSubConn(sc) // Is this right? Should we really be intercepting this and checking if current or pending? I feel like this will be incorrect on Close() calls
```

Comment:
How do you think this can go wrong on Close()? This should be fine since both functions are protected by the `Mu`.

You will need to remove the sc from map `gsb.scToSubBalancer`

Reply:
I.e. Mutex lock in close, grab pointer to it/clear, release lock, close using local var pointer, calls into here, that balancer is no longer current of pending and this RemoveSubConn call never makes it's way to the ClientConn.

Reply:
After "close using local var pointer", all the SubConns belong to this closed balancer should all be removed. So this is already handled.

And, You will need to remove the sc from map `gsb.scToSubBalancer` in this function

Reply:
Done.